refactor(firmware-binding): add message-index field to all can messages and use a singleton to generate unique index values

ryanthecoder opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
https://opentrons.atlassian.net/browse/RET-1244
https://opentrons.atlassian.net/browse/RET-1245
This is the first phase for connecting full error reporting from firmware to the hardware-controller
This allows the firmware to know and indicate exactly which message caused an issue. this will also facilitate catching any unacknowledged messages which would indicate a communication failure

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

ryanthecoder committed:
refactor(firmware-binding): create a base class message for can messages, also make an acknowledgement message for responses

ryanthecoder committed:
feature(firmware-bindings): add a singlton to generate unique message indexes

ryanthecoder committed:
feature(firmware-bindings): add a test that makes sure message index is unique for each message

ryanthecoder committed:
fix(firmware-bindings): override __eq__ for messages so that we can exclude message_index. otherwise all the tests fail.

ryanthecoder committed:
fix(firmware-bindings): fix formating and liniting issues

ryanthecoder committed:
fix(firmware-bindings): update the dataclass hierarchy to make lint AND testing happy

sfoster1 reviewed (approved):
Looks good, but let's hold off on merging until the fw side is ready - looks like it'll break comms generally until that happens

ryanthecoder committed:
fix(firmware-binding) the message_index should be part of the payload not the message

ryanthecoder committed:
remove eq stuff that isn't needed anymore

ryanthecoder committed:
fix(firmware-bindings): update the eq for payload, and make sure the message_index is a field of the payload.
since we don't have the kw_only option for dataclass field we had to add some hacks around it

ryanthecoder committed:
fix(firmware-binding): ignore some attr errors that are already protected by other logic

ryanthecoder committed:
remove unnecessary post_init

ryanthecoder committed:
remove unneedded lint exception

ryanthecoder committed:
add acknowledgement to message def list

ryanthecoder committed:
fix limit_switches.py

ryanthecoder committed:
don't overwrite message index

ryanthecoder committed:
fix can_mon for python3.10 (for emulation)

ryanthecoder committed:
debugging

ryanthecoder committed:
fixups

ryanthecoder committed:
ahhh

ryanthecoder committed:
cleanup

ryanthecoder committed:
format

ryanthecoder committed:
make message index default to none and test against than when seeing to overrride or not

ryanthecoder committed:
fix the downloader messages checksum calculation and fix the tests

ryanthecoder committed:
format

ryanthecoder committed:
make lint happy

ryanthecoder committed:
fix d-word alignment requirement

ryanthecoder committed:
add a guard against invalid data sizes

ryanthecoder committed:
fix(fw_update): make sure fw_update_data address argument is doubleword aligned

ryanthecoder commented:
Going to reopen this PR with a stable branch containing this an other PRs
